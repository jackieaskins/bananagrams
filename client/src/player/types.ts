export type Player = {
  userId: string;
  username: string;
  isReady: boolean;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
};
