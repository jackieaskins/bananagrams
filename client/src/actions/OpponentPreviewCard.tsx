import { useMemo } from 'react';

import DraggableCard from '../card/DraggableCard';
import {
  useGameBoards,
  useGameHands,
  useGamePlayers,
} from '../game/stateHooks';
import PreviewCarousel from '../preview/PreviewCarousel';
import { getUserId } from '../socket';

type OpponentPreviewCardProps = {
  visible: boolean;
};

const OpponentPreviewCard = ({
  visible,
}: OpponentPreviewCardProps): JSX.Element | null => {
  const currentUserId = getUserId();
  const players = useGamePlayers();
  const boards = useGameBoards();
  const hands = useGameHands();

  const opponents = useMemo(
    () => players.filter(({ userId }) => userId !== currentUserId),
    [currentUserId, players]
  );

  return visible ? (
    <DraggableCard
      cardCSS={{
        width: '300px',
        height: '300px',
        marginTop: '5px',
      }}
      bodyStyle={{ height: '100%' }}
    >
      <PreviewCarousel players={opponents} boards={boards} hands={hands} />
    </DraggableCard>
  ) : null;
};

export default OpponentPreviewCard;
