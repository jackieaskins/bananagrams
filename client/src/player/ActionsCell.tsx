import { DeleteOutlined } from '@ant-design/icons';
import { Button } from 'antd';

import { useCurrentPlayer } from '../game/stateHooks';
import { kickPlayer } from '../socket';
import { Player } from './types';

type ActionsCellProps = {
  player: Player;
};

const ActionsCell = ({
  player: { userId },
}: ActionsCellProps): JSX.Element | null => {
  const currentPlayer = useCurrentPlayer();

  if (!currentPlayer) {
    return null;
  }

  const { isAdmin, userId: currentPlayerId } = currentPlayer;

  if (isAdmin && userId !== currentPlayerId) {
    return (
      <Button
        icon={<DeleteOutlined />}
        onClick={() => {
          kickPlayer({ userId });
        }}
        type="text"
      />
    );
  }

  return null;
};

export default ActionsCell;
