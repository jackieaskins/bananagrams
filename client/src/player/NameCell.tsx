import { CrownOutlined, KeyOutlined } from '@ant-design/icons';
import { Space } from 'antd';

import { Player } from '../players/types';

type NameCellProps = {
  player: Player;
};

const NameCell = ({
  player: { isAdmin, isTopBanana, username },
}: NameCellProps): JSX.Element => (
  <Space>
    <span>{username}</span>
    {isAdmin && (
      <span>
        <KeyOutlined />
      </span>
    )}
    {isTopBanana && (
      <span>
        <CrownOutlined />
      </span>
    )}
  </Space>
);

export default NameCell;
