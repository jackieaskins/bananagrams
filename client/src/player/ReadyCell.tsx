import { CheckOutlined, CloseOutlined } from '@ant-design/icons';
import { Checkbox } from 'antd';

import { getUserId, setReady } from '../socket';
import { Player } from './types';

type ReadyCellProps = {
  player: Player;
};

const ReadyCell = ({
  player: { userId, isReady },
}: ReadyCellProps): JSX.Element => {
  const isCurrentUser = userId === getUserId();

  if (isCurrentUser) {
    return (
      <Checkbox
        checked={isReady}
        onChange={({ target: { checked } }) => {
          setReady({ isReady: checked });
        }}
      />
    );
  }

  if (isReady) {
    return <CheckOutlined css={{ color: 'green' }} />;
  }

  return <CloseOutlined css={{ color: 'red' }} />;
};
export default ReadyCell;
