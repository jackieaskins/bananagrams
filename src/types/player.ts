import { Board } from "./board";
import { Hand } from "./hand";

export enum PlayerStatus {
  NOT_READY = "NOT_READY",
  SPECTATING = "SPECTATING",
  READY = "READY",
}

export type Player = {
  userId: string;
  username: string;
  status: PlayerStatus;
  isTopBanana: boolean;
  isAdmin: boolean;
  gamesWon: number;
  hand: Hand;
  board: Board;
};
