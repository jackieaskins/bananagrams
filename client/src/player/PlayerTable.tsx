import { Table, Typography } from 'antd';
import { ColumnType } from 'antd/lib/table';

import { useCurrentPlayer, useGamePlayers } from '../game/stateHooks';
import ActionsCell from './ActionsCell';
import NameCell from './NameCell';
import ReadyCell from './ReadyCell';
import { Player } from './types';

const COLUMN_ALIGN = 'center';
const MAX_PLAYERS = 8;

const { Text } = Typography;

const actionsColumn: ColumnType<Player> = {
  align: COLUMN_ALIGN,
  key: 'actions',
  render: function ActionsRender(_: any, player: Player) {
    return <ActionsCell player={player} />;
  },
};

const PlayerTable = (): JSX.Element => {
  const players = useGamePlayers();
  const isAdmin = !!useCurrentPlayer()?.isAdmin;

  return (
    <Table
      bordered
      columns={[
        {
          title: 'Players',
          children: [
            {
              align: COLUMN_ALIGN,
              key: 'ready',
              title: 'Ready?',
              render: function ReadyRender(_, player) {
                return <ReadyCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              key: 'player',
              title: 'Player',
              render: function PlayerRender(_, player) {
                return <NameCell player={player} />;
              },
            },
            {
              align: COLUMN_ALIGN,
              dataIndex: 'gamesWon',
              title: 'Games won',
            },
            ...(isAdmin && players.length > 1 ? [actionsColumn] : []),
          ],
        },
      ]}
      dataSource={players}
      footer={() => (
        <div css={{ fontSize: '12px', textAlign: 'center' }}>
          <Text type="secondary">
            The game will start when all players are ready
            <br />
            Current player count: {players.length} / {MAX_PLAYERS}
          </Text>
        </div>
      )}
      pagination={false}
      rowKey="userId"
      size="middle"
    />
  );
};

export default PlayerTable;
