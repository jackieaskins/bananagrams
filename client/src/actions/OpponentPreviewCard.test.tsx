import { shallow } from 'enzyme';

import { playerFixture } from '../fixtures/player';
import OpponentPreviewCard from './OpponentPreviewCard';

jest.mock('../game/stateHooks', () => ({
  useGameBoards: jest.fn(() => ({ currentPlayer: {}, opponent: {} })),
  useGameHands: jest.fn(() => ({ currentPlayer: [], opponent: [] })),
  useGamePlayers: jest.fn(() => [
    playerFixture({ userId: 'currentPlayer' }),
    playerFixture({ userId: 'opponent' }),
  ]),
}));

jest.mock('../socket', () => ({
  getUserId: jest.fn(() => 'currentPlayer'),
}));

describe('<OpponentPreviewCard />', () => {
  const renderComponent = (propOverrides = {}) =>
    shallow(<OpponentPreviewCard visible {...propOverrides} />);

  it('renders draggable card with preview carousel if visible', () => {
    expect(renderComponent({ visible: true })).toMatchSnapshot();
  });

  it('does not render if not visible', () => {
    expect(renderComponent({ visible: false })).toMatchSnapshot();
  });
});
