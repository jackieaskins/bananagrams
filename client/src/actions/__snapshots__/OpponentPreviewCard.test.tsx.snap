// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OpponentPreviewCard /> does not render if not visible 1`] = `""`;

exports[`<OpponentPreviewCard /> renders draggable card with preview carousel if visible 1`] = `
<DraggableCard
  cardCSS={
    Object {
      "marginTop": "5px",
      "width": "300px",
    }
  }
>
  <Memo(PreviewCarousel)
    boards={
      Object {
        "currentPlayer": Object {},
        "opponent": Object {},
      }
    }
    hands={
      Object {
        "currentPlayer": Array [],
        "opponent": Array [],
      }
    }
    players={
      Array [
        Object {
          "gamesWon": 0,
          "isAdmin": false,
          "isReady": false,
          "isTopBanana": false,
          "userId": "opponent",
          "username": "username",
        },
      ]
    }
  />
</DraggableCard>
`;
